pkg/controller/account/iam.go

```diff
@@ -0,0 +1,185 @@
+
+	GetFederationTokenInput := sts.GetFederationTokenInput{
+		DurationSeconds: DurationSeconds,
+		Name:            FederatedUserName,
+		PolicyArns:      PolicyArns,
+	}
+
+	// Get Federated token credentials to build console URL
+	GetFederationTokenOutput, err := awsclient.GetFederationToken(&GetFederationTokenInput)
+
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok {
+			// Get error details
+			reqLogger.Error(err, fmt.Sprintf("Error: %s, %s", awsErr.Code(), awsErr.Message()))
+			return "", err
+		}
+
+		return "", err
+	}
+	// Build JSON credentials for federation requets
+	federationCredentials := map[string]string{
+		"sessionId":    *GetFederationTokenOutput.Credentials.AccessKeyId,
+		"sessionKey":   *GetFederationTokenOutput.Credentials.SecretAccessKey,
+		"sessionToken": *GetFederationTokenOutput.Credentials.SessionToken,
+	}
+
+	jsonCredentials, err := json.Marshal(federationCredentials)
+
+	// TODO better error here
+	if err != nil {
+		fmt.Println("Error serializing json")
+		fmt.Println(err)
+		return "", err
+	}
+
+	// Build URL to request Signin Token via Federation end point
+	baseFederationURL, err := url.Parse(federationEndPointURL)
+
+	if err != nil {
+		fmt.Println("Malformed URL: ", err.Error())
+		return "", err
+	}
+
+	federationParams := url.Values{}
+
+	federationParams.Add("Action", "getSigninToken")
+	federationParams.Add("SessionType", "json")
+	federationParams.Add("Session", string(jsonCredentials))
+
+	baseFederationURL.RawQuery = federationParams.Encode()
+
+	// Make HTTP request to retrieve Federated Signin Token
+	res, err := http.Get(baseFederationURL.String())
+
+	if err != nil {
+		fmt.Printf("Error requesting Signin Token from: %s\n", baseFederationURL.String())
+		return "", err
+	}
+
+	defer res.Body.Close()
+
+	body, err := ioutil.ReadAll(res.Body)
+
+	if err != nil {
+		fmt.Printf("Unable to read response body: %s", baseFederationURL.String())
+		return "", err
+	}
+
+	var SigninResponse awsSigninTokenRepsonse
+
+	// Unmarshal JSON response so we can extract the signin token
+	err = json.Unmarshal(body, &SigninResponse)
+
+	if err != nil {
+		fmt.Printf("Error unmarshalling Federated Signin Response JSON")
+		return "", err
+	}
+
+	signinFederationURL, err := url.Parse(federationEndPointURL)
+	if err != nil {
+		fmt.Println("Malformed URL: ", err.Error())
+	}
+
+	signinParams := url.Values{}
+
+	signinParams.Add("Action", "login")
+	signinParams.Add("Destination", awsConsoleURL)
+	signinParams.Add("Issuer", issuer)
+	signinParams.Add("SigninToken", SigninResponse.SigninToken)
+
+	signinFederationURL.RawQuery = signinParams.Encode()
+
+	// Return Signin Token
+	return signinFederationURL.String(), nil
+
+}
+
+// BuildSTSUser takes all parameters required to create a user, user secret
+func (r *ReconcileAccount) BuildSTSUser(reqLogger logr.Logger, awsSetupClient awsclient.Client, awsClient awsclient.Client, account *awsv1alpha1.Account, accountID string, nameSpace string) (string, error) {
```

Comment:
Does this function need to take in both an Account and AccountID? Would there be an instance where these would be different ?

Reply:
Good pickup, not sure why I had both.